src/components/MenuList/index.tsx

```diff
@@ -30,124 +31,195 @@ export const MenuList = React.forwardRef<HTMLDivElement, MenuListProps>(
+        const item = controller.itemsList.current[index];
+
+        if (e.key === 'ArrowUp') {
+          controller.focusPrevious();
+        } else if (e.key === 'ArrowDown') {
+          controller.focusNext();
+        } else if (!item?.hasSubmenu && e.key === 'Enter') {
+          item.onSelect?.(false);
+          root.toggle(false);
+        } else if (e.key === 'ArrowRight' || e.key === 'Enter') {
+          controller.requestSubmenu(controller.getFocusedIndex());
+        } else if (e.key === 'ArrowLeft' || e.key === 'Escape') {
+          if (parentController) {
+            parentController.hideSubmenu();
+          } else if (e.key === 'Escape') {
+            root.toggle(false);
+          }
+        } else {
+          controller.focusUsingText(e.key);
+        }
+      },
+      [root, containerRef, controller, parentController],
+    );
 
-      const popup = list.popup.current;
+    const onWheel = React.useCallback(
+      (e: React.WheelEvent) => {
+        if (!containerRef.current) return;
 
-      if (popup) {
-        ref.style.maxWidth = popup.maxWidth + 'px';
-        ref.style.maxHeight = popup.maxHeight - MENU_LIST_PADDING_Y + 'px';
-        containerRef.style.maxHeight = popup.maxHeight + 'px';
+        e.stopPropagation();
 
-        blurRef.current.style.height =
-          containerRef.clientHeight + MENU_LIST_PADDING_Y * 2 + 'px';
+        containerRef.current.scrollTop =
+          containerRef.current.scrollTop + e.deltaY;
+      },
+      [containerRef],
+    );
 
-        setPosition(list.ref.current, popup.x, popup.y);
+    React.useEffect(() => {
+      if (parentController) {
+        controller.ref.current?.focus();
       }
-    }, [menu, list, x, y, parentWidth, parentHeight]);
+    }, [parentController, controller?.ref]);
 
-    const { onMouseDown: _onMouseDown } = useDisableScrollButton();
-
-    return (
+    const el = (
       <StyledMenuList
-        ref={mergeRefs(list.ref, ref)}
+        ref={controller.ref}
         tabIndex={-1}
-        onKeyDown={mergeEvents(onKeyDown, list.props.onKeyDown)}
-        onWheel={mergeEvents(onWheel, list.props.onWheel)}
-        onBlur={mergeEvents(onBlur, list.props.onBlur)}
-        onMouseDown={mergeEvents(onMouseDown, _onMouseDown)}
-        isOpen={menu?.isOpen}
+        isOpen={root?.isOpen}
+        onKeyDown={onKeyDown}
+        onBlur={onBlur}
+        onWheel={onWheel}
         {...props}
       >
         <BlurEffect ref={blurRef} />
-        {menu?.isOpen && (
-          <Container ref={list.containerRef}>
-            <MenuListContext.Provider value={list}>
+        {root?.isOpen && (
+          <Container ref={containerRef}>
+            <MenuListContext.Provider value={controller}>
               {children}
             </MenuListContext.Provider>
           </Container>
         )}
       </StyledMenuList>
     );
+
+    return parentController == null
+      ? createPortal(el, document.getElementById('wexond-ui-menu-portal')!)
```

Comment:
This probably should be handled by embedders (i.e. developer should decide whether to use a portal).